validator/src/validator.rs

```diff
@@ -447,10 +447,12 @@ where
     /// Check and update if we can enforce the tx validity window.
     /// This is important if we use the state sync and do not have the relevant parts of the history yet.
     fn update_can_enforce_validity_window(&mut self) {
-        if !self.blockchain_state.can_enforce_validity_window
-            && self.blockchain.read().can_enforce_validity_window()
-        {
-            self.blockchain_state.can_enforce_validity_window = true;
+        let old_can_enforce_validity_window = self.blockchain_state.can_enforce_validity_window;
+        self.blockchain_state.can_enforce_validity_window =
+            self.blockchain.read().can_enforce_validity_window();
+
+        // Re-initialize validator when flag returns to true.
+        if !old_can_enforce_validity_window && self.blockchain_state.can_enforce_validity_window {
```

Comment:
I understand that caching `can_enforce_validity_window` was an optimization to not have to take a read lock on the blockchain and call a method for _every single block_. This change is now made so that the state can change from `true` at the launch of a fresh node back to `false` after sync, and then to `true` at the next election block.

Since the underlying condition of history store verifying-capability only changes on election blocks, maybe this state can still be cached during epochs, and the method still only called on epoch changes? That would correctly set it to `false` after sync, and back to `true` after the next election block.

I'm talking about full nodes here, but it would also apply to history nodes I believe.

Reply:
The flag can change only on macro blocks, so we improved the caching a bit.